File: August/27 08 25/removenode.ex
defmodule RemoveNth do
  # Build linked list as tuples {val, next}
  def build_list([]), do: nil
  def build_list([h | t]), do: {h, build_list(t)}

  # Print linked list
  def print_list(nil), do: IO.puts("")
  def print_list({val, next}) do
    IO.write("#{val} ")
    print_list(next)
  end

  # Main function: remove nth node from end
  def remove_nth_from_end(head, n) do
    {_len, new_head} = helper(head, n)
    new_head
  end

  # Helper: returns {length, rebuilt list}
  defp helper(nil, _n), do: {0, nil}
  defp helper({val, next}, n) do
    {len, rebuilt} = helper(next, n)
    curr_len = len + 1

    if curr_len == n do
      {curr_len, rebuilt}  # skip current node
    else
      {curr_len, {val, rebuilt}}
    end
  end
end

# Example Usage
head = RemoveNth.build_list([1, 2, 3, 4, 5])
new_head = RemoveNth.remove_nth_from_end(head, 2)
RemoveNth.print_list(new_head)  # Output: 1 2 3 5

File: August/27/removenode.erl
%% Linked list as a tuple {Val, Next}, or [] for empty
-module(remove_nth).
-export([remove_nth_from_end/2, build_list/1, print_list/1]).

%% Build a linked list from a normal list
build_list([]) -> [];
build_list([H|T]) -> {H, build_list(T)}.

%% Print linked list
print_list([]) -> io:format("~n");
print_list({Val, Next}) ->
    io:format("~p ", [Val]),
    print_list(Next).

%% Main function
remove_nth_from_end(Head, N) ->
    {_NewHead, _Len, Result} = helper(Head, N),
    Result.

%% Helper: returns {current node, length, rebuilt list}
helper([], _N) -> {[], 0, []};
helper({Val, Next}, N) ->
    {_, Len, Rebuilt} = helper(Next, N),
    CurrLen = Len + 1,
    case CurrLen =:= N of
        true -> {Val, Next, Rebuilt};       %% skip this node
        false -> {Val, Next, {Val, Rebuilt}}
    end.
